### PR TITLE
Add Authentication Metadata Extension

### PR DESCRIPTION
**Pull request by rwinch**
Fixes gh-272

### CODE REVIEW

File: Extensions/Security/Basic.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+     0                   1                   2                   3
+     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    |Username Length|     Username  |    Password              ...
+    +---------------+-----------------------------------------------+
+```
+
+* **Username Length**: Username Length in bytes.
+* **Username**:  The UTF-8 encoded username.  The string MUST NOT be null terminated.
+* **Password**:  The UTF-8 encoded password.  The string MUST NOT be null terminated.
```
Comment (spencergibb) [inline Extensions/Security/Basic.md]: Password Length?

Comment (linux-china) [reply Extensions/Security/Basic.md]: How about tagging for username and password?  https://github.com/rsocket/rsocket/blob/master/Extensions/Routing.md .  The first tag is user name, and the second is password?

Comment (rwinch) [reply Extensions/Security/Basic.md]: > Password Length?

Isn't that just the remaining bytes?

**Comment by OlegDokuka** [comment]
I'm thinking will not be it simpler to having something similar to what there is in HTTP, e.g having a type of auth as a part of metadata, which limits the number of possible metadata types to a minimum and does not require further extending to the extension. 

for example: 

```
     0                   1                   2                   3
     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |Auth ID/Length |         Auth Type                            ...
    +---------------+-----------------------------------------------+
	|   					Encoded Credentials					   ...
    +---------------------------------------------------------------+
```

where Auth ID could be additionally listed (as for WellKnown MimeTypes) or it could be any Auth Type encoded as a string

**Comment by OlegDokuka** [comment]
Hey, @rwinch @spencergibb 

Do you have any updates?

Regards,
Oleh

**Comment by rwinch** [comment]
@OlegDokuka and I spoke about the changes offline. The goal is to avoid having a metadata extension for each type of authentication. However, I believe we should avoid using constructs that work around limitations in HTTP (i.e. `base64(username:password)` in RSocket.

We have defined something similar to Composite Metadata Extension. However, it is specific to Authentication. This allows for custom encoding of the credentials without the need for a custom mime type.

I have pushed the changes for review.

**Comment by OlegDokuka** [comment]
@linux-china @spencergibb @rdegnan @robertroeser @nebhale @stevegury, @yschimke  folks, do you have any comments/objections?

**Comment by yschimke** [comment]
No strong objections. Happy for those using it to prove it in v0 and burn a version if there are mistakes. 

General concerns: doesn’t cover challenges from HTTP and not sure the optimisation is worth having to invent our own non standard encodings.

File: Extensions/Security/Authentication.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+     0                   1                   2                   3
+     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    |A| Auth ID/Len |   Authentication Type                        ...
+    +---------------+---------------+---------------+---------------+
+    |                     Authentication Payload                   ...
+    +---------------+-----------------------------------------------+
+```
+
+* (**A**)uthentication Type: Authentication type is a well known value represented by a unique integer.
```
Comment (stevegury) [inline Extensions/Security/Authentication.md]: It's not clear which bit value does what.
Maybe we should be more explicit, i.e. `0 means Auth Length, 1 means Auth Id`.

**Comment by rwinch** [comment]
@yschimke 

> doesn’t cover challenges from HTTP

I'm sorry. I don't understand. Can you please expand on what you mean by this?

> not sure the optimisation is worth having to invent our own non standard encodings.

How would you encode it then? 

The way a username/password is encoded when using HTTP are based on context. HTTP Basic is base64 encoded because it makes sense for HTTP headers. When a username/password is passed in the body of the request as form parameters (I'd argue this is the most common way username/password is submitted with HTTP), they are URL encoded because that is what makes sense for a form submission. If credentials are submitted as JSON, then they are encoded yet another way.

In short, the way a username/password is encoded when using HTTP is based upon the context. In the same manner, when using RSocket a username/password should be encoded in a way that makes sense for RSocket.

**Comment by yschimke** [comment]
> I'm sorry. I don't understand. Can you please expand on what you mean by this?

This spec is obviously inspired by existing HTTP standards but doesn't consider the full flow of challenge response. 

Instead its focus on prior knowledge usage. 

https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication

**Comment by yschimke** [comment]
> In short, the way a username/password is encoded when using HTTP is based upon the context.

That's not true at all. There is a spec for basic auth including an encoding. 

https://en.m.wikipedia.org/wiki/Basic_access_authentication

**Comment by rwinch** [comment]
> This spec is obviously inspired by existing HTTP standards but doesn't consider the full flow of challenge response.

Are you pointing out that RSocket does not define how to request a user provide credentials? If so, I agree.

@OlegDokuka and I discussed this offline and I believe he was planning on adding a separate ticket to discuss how to handle errors for authentication required, invalid authentication, and not authorized.

I personally felt that including credentials and credential error handling would be separate specifications that eventually link to one another. This would allow for reusing the error handling when including the credentials in a different way. If we want to combine this into a single specification, I'd be open to discussing that. Just let me know.

> That's not true at all for basic auth. There is a spec for basic auth including an encoding.

HTTP Basic Auth defines an encoding, but it also defines the context - an HTTP header field. Username/password included in a form post, JSON, etc do not use the same encoding as HTTP Basic Auth because they are not included in the header. Instead, they use the appropriate encoding for their contexts. Just as a form submission does not use HTTP basic encoding, RSocket should use an encoding that makes sense for the context of metadata.

> Maybe just rename it to username and password instead of basic.

I agree the name basic is not a good choice due to the fact that many will confuse it with HTTP Basic. My hope is that this is differentiated by HTTP Basic Auth vs RSocket basic, but this is still not as clear as I'd prefer.

I'm personally not a fan of username and password as it seems pretty verbose. Perhaps I am alone with this though. If anyone has thoughts on naming, I'd like to hear it.

File: Extensions/Security/Authentication.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# Authentication Extension
+
+_This extension specification is currently incubating.  While incubating the version is 0._
+
+## Introduction
+Authentication is a necessary component to any real world application. This extension specification provides a standardized mechanism for including both the type of credentials and the credentials in metadata payloads.
+
+## Metadata Payload
+This metadata type can be used in a per connection or per stream, and not individual payloads and as such it **MUST** only be used in frame types used to initiate interactions and payloads.  This includes [`SETUP`][s], [`REQUEST_FNF`][rf], [`REQUEST_RESPONSE`][rr], [`REQUEST_STREAM`][rs], [`REQUEST_CHANNEL`][rc], and [`PAYLOAD`][p].  The Metadata MIME Type is `message/x.rsocket.authentication.v0`.
```
[the remark this answers is not in the document]
Comment (OlegDokuka) [reply Extensions/Security/Authentication.md]: Looks like it is a common part of every extension definition -> for instance https://github.com/rsocket/rsocket/blob/master/Extensions/Routing.md#metadata-payload

**Comment by OlegDokuka** [comment]
> @OlegDokuka and I discussed this offline and I believe he was planning on adding a separate ticket to discuss how to handle errors for authentication required, invalid authentication, and not authorized.

yes, I’m planning to rise another ticket for that

> I'm personally not a fan of username and password as it seems pretty verbose. Perhaps I am alone with this though. Perhaps we could use simple? If anyone has thoughts on naming, I'd like to hear it.

Yeah, I like `simple` for that purpose

File: Extensions/Security/Authentication.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (rwinch) [reply Extensions/Security/Authentication.md]: @yschimke You are right that we should not list `PAYLOAD` here because it is not a start of a stream or connection. I pushed https://github.com/rsocket/rsocket/pull/275/commits/fa1a80b9212816b2efce0c538d06ba5366d1a32b to address this.

I kept the commit separate, so it was easier to see the changes. I can clean up the commits later if desired.

File: Extensions/Security/Authentication.md (hunk 2 above)
Comment (rwinch) [reply Extensions/Security/Authentication.md]: @stevegury Thanks for the feedback. I tried to follow the conventions I found in the existing specs. The clarification is on the next line just as it is done in [Composite Metadata](https://github.com/rsocket/rsocket/blob/7278d1dd7963893be77069bb3b699489d65f42ac/Extensions/CompositeMetadata.md#metadata-contents):

> If A flag is set, indicates a [Well-known Auth Type ID][wk].  If A flag is not set, indicates the Authentication Type Length in bytes.

I pushed a fix in https://github.com/rsocket/rsocket/pull/275/commits/de71eaac27ee5b9d231515a9662eed020c8b09a3 I kept the commit separate, so it was easier to see the changes. I can clean up the commits later if desired.

If we decide to stick with these changes, I'd suggest we also update Composite Metadata.

**Comment by rwinch** [comment]
> Yeah, I like simple for that purpose

I pushed a change for this https://github.com/rsocket/rsocket/pull/275/commits/732ad717e6f818ab7bed00594038cf16c630bc1a I kept the commit separate, so it was easier to see the changes. I can clean up the commits later if desired.

**Comment by yschimke** [comment]
To be clear. No intention from me to block forward progress. Just feedback to consider. Happy to ship this. Usage trumps drive by opinions :)

File: Extensions/Security/Authentication.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    |A| Auth ID/Len |   Authentication Type                        ...
+    +---------------+---------------+---------------+---------------+
+    |                     Authentication Payload                   ...
+    +---------------+-----------------------------------------------+
+```
+
+* (**A**)uthentication Type: Authentication type is a well known value represented by a unique integer.  If A flag is set (a value of `1`), indicates a [Well-known Auth Type ID][wk].  If A flag is not set (a value of `0`), indicates the Authentication Type Length in bytes.
+* **Auth ID/Length**: (7 bits = max value 2^7 = 128) Unsigned 7-bit integer.  If A flag is set (a value of `1`), indicates a [Well-known Auth Type ID][wk].  If A flag is not set (a value of `0`), indicates the Authentication Type Length in bytes.
```
Comment (OlegDokuka) [inline Extensions/Security/Authentication.md]:
```suggestion
* **Auth ID/Length**: (7 bits = max value 2^7 = 128) Unsigned 7-bit integer.  Applicable only if A flag is not set (a value of `0`).
```
